File: vscode/src/model/reviewMarkdown.ts
import { ReviewComment, ReviewFile, ReviewState, ReviewStatus } from "./reviewModel";

const FILE_HEADER = /^## (.+)$/;
const COMMENT_LINE = /^- id:([0-9a-fA-F-]+)(?: line:(\d+)(?:-(\d+))?)?(?: cat:([QR]))? (.*)$/;

function escapeBody(body: string): string {
  const normalized = body.replace(/\r\n/g, "\n").replace(/\r/g, "\n");
  let out = "";
  for (const c of normalized) {
    if (c === "\\") out += "\\\\";
    else if (c === "\n") out += "\\n";
    else out += c;
  }
  return out;
}

function unescapeBody(raw: string): string {
  let out = "", i = 0;
  while (i < raw.length) {
    const c = raw[i];
    if (c === "\\" && i + 1 < raw.length) {
      const next = raw[i + 1];
      if (next === "n") { out += "\n"; i += 2; }
      else if (next === "\\") { out += "\\"; i += 2; }
      else { out += c; i += 1; }
    } else { out += c; i += 1; }
  }
  return out;
}

export function parseReviewMarkdown(text: string): ReviewState {
  const lines = text.split("\n");
  if (lines.length < 3 || lines[0].trim() !== "---") return { status: "open", files: [] };
  const statusLine = lines[1].trim();
  const status: ReviewStatus | null =
    statusLine === "status: submitted" ? "submitted" :
    statusLine === "status: open" ? "open" : null;
  if (!status || lines[2].trim() !== "---") return { status: "open", files: [] };

  const files: ReviewFile[] = [];
  let currentFile: string | null = null;
  let currentComments: ReviewComment[] = [];
  const flush = () => { if (currentFile && currentComments.length > 0) files.push({ file: currentFile, comments: currentComments }); };

  for (const line of lines.slice(3)) {
    const header = line.match(FILE_HEADER);
    if (header) { flush(); currentFile = header[1]; currentComments = []; continue; }
    const m = line.match(COMMENT_LINE);
    if (!m) continue;
    currentComments.push({
      id: m[1],
      ...(m[2] ? { line: parseInt(m[2], 10) } : {}),
      ...(m[3] ? { lineEnd: parseInt(m[3], 10) } : {}),
      category: m[4] === "Q" ? "Q" : "R",
      body: unescapeBody(m[5]),
    });
  }
  flush();
  return { status, files };
}

export function renderReviewMarkdown(state: ReviewState): string {
  let out = `---\nstatus: ${state.status === "submitted" ? "submitted" : "open"}\n---\n`;
  for (const file of state.files) {
    if (file.comments.length === 0) continue;
    out += `\n## ${file.file}\n\n`;
    for (const c of file.comments) {
      const lineToken = c.line == null ? "" : c.lineEnd != null ? ` line:${c.line}-${c.lineEnd}` : ` line:${c.line}`;
      const catToken = c.category === "Q" ? " cat:Q" : " cat:R";
      out += `- id:${c.id}${lineToken}${catToken} ${escapeBody(c.body)}\n`;
    }
  }
  return out;
}

File: intellij/src/main/kotlin/io/github/ompreview/model/ReviewMarkdown.kt
package io.github.ompreview.model

// Mirrors the FILE_HEADER / COMMENT_LINE format in omp/src/index.ts — keep both in sync.
//
//   ---
//   status: open|submitted
//   ---
//
//   ## <file>
//
//   - id:<uuid> line:<n>[-<m>] <body>
//   - id:<uuid> <body>                ← file-level (no line: token)

private val FILE_HEADER = Regex("^## (.+)$")
private val COMMENT_LINE = Regex("""^- id:([0-9a-fA-F-]+)(?: line:(\d+)(?:-(\d+))?)?(?: cat:([QR]))? (.*)$""")

// Preserves multi-line comment bodies (the Add Comment / Add File Comment dialogs use a
// multi-line text area, Step 5) on a single physical review.md line: backslash -> "\\",
// a real newline -> the two characters "\n". Mirrors unescapeBody in omp/src/index.ts —
// TS only ever reads this file, so it implements the reverse direction only.
private fun escapeBody(body: String): String {
    val sb = StringBuilder()
    for (c in body.replace("\r\n", "\n").replace('\r', '\n')) {
        when (c) {
            '\\' -> sb.append("\\\\")
            '\n' -> sb.append("\\n")
            else -> sb.append(c)
        }
    }
    return sb.toString()
}

private fun unescapeBody(raw: String): String {
    val sb = StringBuilder()
    var i = 0
    while (i < raw.length) {
        val c = raw[i]
        if (c == '\\' && i + 1 < raw.length) {
            when (raw[i + 1]) {
                'n'  -> { sb.append('\n'); i += 2 }
                '\\' -> { sb.append('\\'); i += 2 }
                else -> { sb.append(c); i += 1 }
            }
        } else {
            sb.append(c); i += 1
        }
    }
    return sb.toString()
}

object ReviewMarkdown {

    fun parse(text: String): ReviewState {
        val lines = text.lines()
        if (lines.size < 3 || lines[0].trim() != "---") return ReviewState()

        val status = when (lines[1].trim()) {
            "status: submitted" -> ReviewStatus.SUBMITTED
            "status: open"      -> ReviewStatus.OPEN
            else                -> return ReviewState()
        }
        if (lines[2].trim() != "---") return ReviewState()

        val files = mutableListOf<ReviewFile>()
        var currentFile: String? = null
        var currentComments = mutableListOf<ReviewComment>()

        fun flush() {
            val file = currentFile ?: return
            if (currentComments.isNotEmpty()) files += ReviewFile(file, currentComments.toList())
        }

        for (line in lines.drop(3)) {
            val header = FILE_HEADER.matchEntire(line)
            if (header != null) {
                flush()
                currentFile = header.groupValues[1]
                currentComments = mutableListOf()
                continue
            }
            val match = COMMENT_LINE.matchEntire(line) ?: continue
            currentComments += ReviewComment(
                id       = match.groupValues[1],
                line     = match.groupValues[2].toIntOrNull(),
                lineEnd  = match.groupValues[3].toIntOrNull(),
                category = when (match.groupValues[4]) {
                    "Q"  -> CommentCategory.QUESTION
                    "R"  -> CommentCategory.REQUEST
                    else -> CommentCategory.REQUEST
                },
                body     = unescapeBody(match.groupValues[5]),
            )
        }
        flush()

        return ReviewState(status, files)
    }

    fun render(state: ReviewState): String {
        val sb = StringBuilder()
        sb.append("---\n")
        sb.append("status: ${if (state.status == ReviewStatus.SUBMITTED) "submitted" else "open"}\n")
        sb.append("---\n")

        for (file in state.files) {
            if (file.comments.isEmpty()) continue
            sb.append("\n## ${file.file}\n\n")
            for (comment in file.comments) {
                val lineToken = when {
                    comment.line == null    -> ""
                    comment.lineEnd != null -> " line:${comment.line}-${comment.lineEnd}"
                    else                    -> " line:${comment.line}"
                }
                val catToken = when (comment.category) {
                    CommentCategory.QUESTION -> " cat:Q"
                    CommentCategory.REQUEST  -> " cat:R"
                }
                val body = escapeBody(comment.body)
                sb.append("- id:${comment.id}$lineToken$catToken $body\n")
            }
        }
        return sb.toString()
    }
}
